### ISSUE TITLE
Differences between memory efficient attention and vanilla eager attention

### ISSUE THREAD

**Issue by nbui-sc** [lucidrains/memory-efficient-attention-pytorch#9]
I'm trying to use your implementation of memory efficient attention (and pure Pytorch FlashAttention implementation as well) to replace the eager implementation of Qwen3 as in this function.

https://github.com/huggingface/transformers/blob/0e1c2817455602d182bd8ebf5fba212e14fb187e/src/transformers/models/qwen3/modeling_qwen3.py#L135

But I noticed that there's a difference between the output of the vanilla eager implementation and your implementation. For one layer, the differences could be up to 1e-2 for one layer.

<img width="1697" height="280" alt="Image" src="https://github.com/user-attachments/assets/fee37ef2-0745-48dd-916f-2e86eaf65669" />

It will be accumulated after multiple layers and eventually lead to a very weird output. I wonder if you have some insights on the place that caused the error? When I debug with a very small input (everything fits one chunk) and check the output at each step, the differences start from the softmax computation.

**Comment by nbui-sc** [comment]
I think the problem is just a precision between bfloat16 and float32. Loading the model with float32 addresses the issue.

**Comment by nbui-sc** [comment]
Btw, this seems to be + instead of - (for autogressive decoding purposes)? 

`q_start_index = ind * q_bucket_size + qk_len_diff`

https://github.com/lucidrains/memory-efficient-attention-pytorch/blob/d54f391370ecbf843a871f0e260425d076995550/memory_efficient_attention_pytorch/flash_attention.py#L64
